File: android-agent/app/src/main/java/com/ppaass/ai/agent/DirectRuleDomains.java
package com.ppaass.ai.agent;

import com.google.common.net.InternetDomainName;
import com.google.common.net.InetAddresses;

import java.util.*;

final class DirectRuleDomains {
    private DirectRuleDomains() {
    }

    static String toDirectRule(String domain) {
        String normalized = normalizeDomain(domain);
        int firstDot = normalized.indexOf('.');
        if (firstDot <= 0) {
            return normalized;
        }

        String parent = normalized.substring(firstDot + 1);
        return isPublicSuffix(parent) ? normalized : "*." + parent;
    }

    static List<String> toDirectRules(Collection<String> domains) {
        LinkedHashSet<String> rules = new LinkedHashSet<>();
        for (String domain : domains) {
            String rule = toDirectRule(domain);
            if (!rule.isEmpty()) {
                rules.add(rule);
            }
        }
        return new ArrayList<>(rules);
    }

    static List<String> toDirectRules(Collection<String> domains, Collection<String> addresses) {
        LinkedHashSet<String> rules = new LinkedHashSet<>(toDirectRules(domains));
        for (String value : addresses) {
            String address = normalizeAddress(value);
            if (!address.isEmpty()) {
                rules.add(address);
            }
        }
        return new ArrayList<>(rules);
    }

    static boolean ruleCoversDomain(String rule, String domain) {
        String normalizedRule = rule == null ? "" : rule.trim().toLowerCase(Locale.US);
        String normalizedDomain = normalizeDomain(domain);
        if (normalizedRule.equals(normalizedDomain)) {
            return true;
        }
        if (!normalizedRule.startsWith("*.")) {
            return false;
        }
        String suffix = normalizedRule.substring(2);
        return !normalizedDomain.equals(suffix) && normalizedDomain.endsWith("." + suffix);
    }

    static List<String> directRulesMatchingDomainsAndAddresses(
            Collection<String> existingRules,
            Collection<String> domains,
            Collection<String> addresses) {
        List<String> normalizedDomains = new ArrayList<>();
        for (String domain : domains) {
            String normalized = normalizeDomain(domain);
            if (!normalized.isEmpty()) {
                normalizedDomains.add(normalized);
            }
        }

        HashSet<String> addressKeys = new HashSet<>();
        for (String value : addresses) {
            String address = normalizeAddress(value);
            if (!address.isEmpty()) {
                addressKeys.add(address);
            }
        }

        List<String> matches = new ArrayList<>();
        for (String rule : existingRules) {
            String normalizedRuleAddress = normalizeAddress(rule);
            boolean matchesAddress = !normalizedRuleAddress.isEmpty()
                    && addressKeys.contains(normalizedRuleAddress);
            boolean matchesDomain = false;
            for (String domain : normalizedDomains) {
                if (ruleCoversDomain(rule, domain)) {
                    matchesDomain = true;
                    break;
                }
            }
            if (matchesAddress || matchesDomain) {
                matches.add(rule);
            }
        }
        return matches;
    }

    private static String normalizeDomain(String domain) {
        String normalized = domain == null ? "" : domain.trim().toLowerCase(Locale.US);
        while (normalized.endsWith(".")) {
            normalized = normalized.substring(0, normalized.length() - 1);
        }
        return normalized;
    }

    private static String normalizeAddress(String value) {
        String address = value == null ? "" : value.trim();
        if (address.isEmpty()) {
            return "";
        }
        try {
            return InetAddresses.toAddrString(InetAddresses.forString(address));
        } catch (IllegalArgumentException ignored) {
            return "";
        }
    }

    private static boolean isPublicSuffix(String domain) {
        try {
            InternetDomainName name = InternetDomainName.from(domain);
            return !name.hasPublicSuffix() || name.isPublicSuffix();
        } catch (IllegalArgumentException ignored) {
            return true;
        }
    }
}

File: desktop-agent-ui/src/directRuleDomains.ts
import { getDomain } from "tldts";

export function domainToDirectRule(domain: string) {
  const normalized = domain.trim().replace(/\.$/, "").toLowerCase();
  const firstDot = normalized.indexOf(".");
  if (firstDot <= 0) {
    return normalized;
  }

  const parent = normalized.slice(firstDot + 1);
  // A wildcard must never target a public suffix such as *.com or *.co.uk.
  return getDomain(parent, { allowPrivateDomains: true }) ? `*.${parent}` : normalized;
}

export function domainsToDirectRules(domains: string[]) {
  const rules = domains.map(domainToDirectRule);
  return [...new Set(rules.filter(Boolean))];
}

export function domainsAndAddressesToDirectRules(domains: string[], addresses: string[]) {
  const rules = domainsToDirectRules(domains);
  for (const value of addresses) {
    const address = value.trim();
    if (isIpAddress(address) && !rules.includes(address)) {
      rules.push(address);
    }
  }
  return rules;
}

export function selectedDomainsToNewDirectRules(
  selectedDomains: string[],
  addresses: string[],
  existingRules: string[]
) {
  const domains = [
    ...new Map(
      selectedDomains
        .map((domain) => domain.trim().replace(/\.$/, ""))
        .filter(Boolean)
        .map((domain) => [domain.toLowerCase(), domain] as const)
    ).values()
  ].filter(
    (domain) => !existingRules.some((rule) => directRuleCoversDomain(rule, domain))
  );
  const existingRuleKeys = new Set(
    existingRules.map((rule) => rule.trim().toLowerCase())
  );

  return domainsAndAddressesToDirectRules(domains, addresses).filter(
    (rule) => !existingRuleKeys.has(rule.trim().toLowerCase())
  );
}

export function isIpAddress(value: string) {
  const candidate = value.trim();
  if (candidate.includes(":")) {
    try {
      return new URL(`http://[${candidate}]/`).hostname.length > 2;
    } catch {
      return false;
    }
  }

  const octets = candidate.split(".");
  return (
    octets.length === 4 &&
    octets.every((octet) => /^\d{1,3}$/.test(octet) && Number(octet) <= 255)
  );
}

export function directRuleCoversDomain(rule: string, domain: string) {
  const normalizedRule = rule.trim().toLowerCase();
  const normalizedDomain = domain.trim().replace(/\.$/, "").toLowerCase();
  if (normalizedRule === normalizedDomain) {
    return true;
  }
  if (!normalizedRule.startsWith("*.")) {
    return false;
  }
  const suffix = normalizedRule.slice(2);
  return normalizedDomain !== suffix && normalizedDomain.endsWith(`.${suffix}`);
}

export function directRulesMatchingDomainsAndAddresses(
  existingRules: string[],
  domains: string[],
  addresses: string[]
) {
  const normalizedDomains = domains
    .map((domain) => domain.trim().replace(/\.$/, ""))
    .filter(Boolean);
  const addressKeys = new Set(
    addresses
      .map((address) => address.trim())
      .filter(isIpAddress)
      .map((address) => address.toLowerCase())
  );

  return existingRules.filter((rule) => {
    const normalizedRule = rule.trim().toLowerCase();
    return (
      addressKeys.has(normalizedRule) ||
      normalizedDomains.some((domain) => directRuleCoversDomain(rule, domain))
    );
  });
}
